Support for new format of CCA top level token to match CMW changes

SimonFrost-Arm opened the pull request:
Support for updated (breaking) change in CCA top level token made to conform with updated CMW definition.
Change to spec can be found in RMM 1.1-alp14.

SimonFrost-Arm committed:
Support for new format of CCA top level token to match CMW changes

SimonFrost-Arm committed:
Support for new format of CCA top level token to match CMW changes

SimonFrost-Arm committed:
feat: Support for updated definition of CCA top level token made to conform with updated CMW definition.
Change to spec can be found in RMM 1.1-alp14.
Code will still decode EAT Collection formatted tokens

Signed-off-by: simfro01 <simon.frost@arm.com>

SimonFrost-Arm committed:
Merge branch 'new-cca-top-token' of github.com:veraison/ccatoken into new-cca-top-tokenSupport for updated (breaking) change in CCA top level token made to conform with updated CMW definition.
Change to spec can be found in RMM 1.1-alp14.

yogeshbdeshpande commented:
The changes LGTM, just one question:

I noticed the RMM spec referred in the README.md still uses
`cca-token = #6.399(cca-token-collection) ;`

Is there a plan to change the spec? as currently the spec seems out of sync with the proposed changes.

SimonFrost-Arm commented:
> Is there a plan to change the spec? as currently the spec seems out of sync with the proposed changes.
The change will be in the upcoming 1.1alp14

yogeshbdeshpande reviewed (approved):
LGTM!

setrofim reviewed (approved):
Please squash the commits, add `Signed-off-by` footer, and update  the commit message to align with [Conventional Commits][1].

Apart from that, LGTM.

[1]: https://www.conventionalcommits.org/en/v1.0.0/

thomas-fossati commented:
I think we can be a bit more conservative on the decoding side.
The fact that the CBOR tags are different allows us to support the EAT collection in parallel with the CMW collection pretty straightforwardly.

Something like:

```golang
 var (
        tags = []encoding.CBORTagEntry{
                {Type: CBORCollection{}, Tag: 399},
                {Type: CMWCollection{}, Tag: 903},
        }
```
and then

```golang
type CMWCollection struct {
	PlatformTokenColl CMWCollectionEntry `cbor:"44234,keyasint"`
	RealmTokenColl    CMWCollectionEntry `cbor:"44241,keyasint"`
```

Or do that manually, by first decoding the top-level tag, and then dispatching to the right collection decoder:
```golang

  var t cbor.RawTag

  if err = t.UnmarshalCBOR(b); err != nil {
          return fmt.Errorf("unmarshal top-level CBOR Tag: %w", err)
  }
  
  switch t.Number {
  case 903: // do new
  case 399: // do legacy
  }
```  
  

This would give senders the ability to align with the new format at their own pace.

SimonFrost-Arm committed:
feat: Support for change in CCA top level token made to conform with updated CMW definition.
Change to spec can be found in RMM 1.1-alp14.
Change supports new token format and will still decode previous EAT Collection format

Signed-off-by: simfro01 <simon.frost@arm.com>

SimonFrost-Arm committed:
feat: Support for change in CCA top level token made to conform with updated CMW definition.
Change to spec can be found in RMM 1.1-alp14.
Change supports new token format and will still decode previous EAT Collection format

Signed-off-by: simfro01 <simon.frost@arm.com>

SimonFrost-Arm commented:
Latest commit adds support for reading old (399) tokens as well as the full CMW tokens

yogeshbdeshpande commented:
> Latest commit adds support for reading old (399) tokens as well as the full CMW tokens

Apologies for a late reply, but wanted to check, is this complexity really required, at this stage supporting two types of collection ?

The products using token type are not yet out in the field, so if we can unilaterally change that is a better fit for now ?
Just my two cents!

thomas-fossati commented:
> > Latest commit adds support for reading old (399) tokens as well as the full CMW tokens
> 
> Apologies for a late reply, but wanted to check, is this complexity really required, at this stage supporting two types of collection ?

Yes, there is running code, in testing labs and in CIs, that’s better not brutally break :-)

SimonFrost-Arm committed:
feat: Support for change in CCA top level token made to conform with updated CMW definition.

Change to spec can be found in RMM 1.1-alp14.
Change supports new token format and will still decode previous EAT Collection format

SimonFrost-Arm commented:
Merged separately to cleanup messy commit history
